ibis/expr/operations/udf.py

```diff
@@ -0,0 +1,157 @@
+from __future__ import annotations
+
+import functools
+import inspect
+import itertools
+import typing
+from typing import TYPE_CHECKING, Any, Callable, TypeVar
+
+from public import public
+
+import ibis.common.exceptions as exc
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.rules as rlz
+from ibis import util
+from ibis.common.annotations import attribute
+from ibis.common.collections import FrozenDict
+
+if TYPE_CHECKING:
+    import ibis.expr.types as ir
+
+
+@public
+class ScalarUDF(ops.Value):
+    @attribute.default
+    def output_shape(self):
+        return rlz.highest_precedence_shape(
+            itertools.chain.from_iterable(
+                util.promote_list(getattr(self, name)) for name in self.argnames
+            )
+        )
+
+
+@public
+class SingleScalarUDF(ScalarUDF):
+    pass
+
+
+@public
+class VectorizedScalarUDF(ScalarUDF):
+    pass
+
+
+def _wrap(wrapper, fn: Callable | None = None, *args: Any, **kwargs: Any) -> Callable:
+    """Wrap a function `fn` with `wrapper`, allowing zero arguments when used as part of a decorator."""
+    if fn is None:
+        return lambda fn: functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+    return functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+
+
+S1 = TypeVar("S1", bound=ScalarUDF)
+S2 = TypeVar("S2", bound=ScalarUDF)
+
+
+class UDF:
```

Comment:
Could this code live in the `builders.py`?

Reply:
Maybe, but I think I can do that in a follow-up too.